Upgrade torch to v2.0

zberkom opened the pull request:
## Brief Description

This PR copies over all the v2.0 code developed in a clean mix project [repo](https://github.com/zberkom/torch). The goal of this PR is to rewrite torch to take advantage of the Phoenix generators, make styling easier, and support the latest version of Phoenix.

## Details

#### Screenshots 

Index:
![image](https://user-images.githubusercontent.com/7085617/36336303-6c83fb18-133b-11e8-8931-f1599d41a6e7.png)

Show: 
![image](https://user-images.githubusercontent.com/7085617/36336336-aadc7d0e-133b-11e8-970e-1f33ac778371.png)

Edit: 
![image](https://user-images.githubusercontent.com/7085617/36336338-b7bdae26-133b-11e8-9a76-bd4cd767e6a7.png)

#### Commits Included from separate Repo

:tada: First commit
:sparkles: Add initial mix gen tasks and templates
:sparkles: Add html generator task
:boom: Add views and update index.html template
:recycle: Updates to html templates
:sparkles: Add pagination function to context
:white_check_mark: Update controller test template
:white_check_mark: Update test cases
:boom: Add initial styles setup
:boom: Add router and update layout.html
:boom: Add theme styling for header & toolbar
:boom: Add styling for index page filters
:boom: Add index table styling
:recycle: Updates to table styling
:boom: Add styles to pagination
:boom: Add show page styles
:boom: Add styles to edit & new
:recycle: Add table responsive styles
:boom: Add javascript
:boom: Add form error styling
:recycle: Refactor all templates to match theme layout
:boom: Add flash message support and make minor updates to templates
:sparkles: Update the README
:pencil: Update readme and add release script
:boom: Add bin/setup script and example phoenix 1.3 app
:white_check_mark: Add tests to example app

zberkom committed:
:boom: Copy over files from v2.0 repo

zberkom committed:
:arrow_down: Downgrade expected elixir version for semaphore

zberkom committed:
:white_check_mark: Update test database creds

zberkom committed:
Update bin/setup script

danielberkompas commented:
Overall, this is looking great! I love how simple this approach is. We can address these comments here or after merge.

#### Mix Tasks

`mix torch.install [project] [template_format]` works a little strangely.

- It adds a `_web` suffix to the project name. So, in an umbrella app, I `cd apps/my_app_web` and then have to run `mix torch.install my_app eex`. I would expect to run `mix torch.install my_app_web eex` or even better, `mix torch.install eex`.

- It installs all the templates for `phx.gen.context` and `phx.gen.html` to the `priv` directory. If I then use `mix torch.gen.html`, it will (unsuccessfully) attempt to remove those templates. The `install` task is fighting the `gen` task. It seems like we should either:

    - Only install the `admin.html` layout, OR
    - Remove the `torch.gen.*` tasks in favor of the `phx.gen` tasks.

`mix torch.gen.html` sometimes fails to remove the templates from the priv directory, because it doesn't look in the right file paths.

#### Customization

- It might be better to brand Torch as a `CRUD` generator, rather than an "Admin" generator. People might not always use it to create an Admin. We might want to make the name of the layout customizable, or perhaps make it `torch.html` instead of `admin.html`.

#### Formatting

- If we keep the `torch.gen.*` tasks, we should run `mix format` after they run to clean up their output.

#### Documentation

- Can we use the logo in the README?
<img width="160" alt="admin" src="https://user-images.githubusercontent.com/1186992/36438765-b54b60a2-161f-11e8-9a2a-13b2c4f24835.png">

- If we keep `torch.gen.*` tasks, we should add instructions to the end of them to add a link to the new page in your `layouts/torch.html` in the `<nav>`

- Can we maybe remove the :fire: emoji from the command line output? It feels a little retro or amateurish to me. That might just be me though.

#### Bugs

- Clicking "X" doesn't close the flash message.
<img width="327" alt="admin" src="https://user-images.githubusercontent.com/1186992/36438977-55581e46-1620-11e8-8efe-6961610f898b.png">

silasjmatson commented on .credo.exs:
```diff
@@ -41,68 +50,94 @@
+        #
+        {Credo.Check.Refactor.ABCSize, false},
+        {Credo.Check.Refactor.AppendSingleItem, false},
+        {Credo.Check.Refactor.VariableRebinding, false},
```
I think we should opt into [VariableRebinding](https://github.com/rrrene/credo/blob/master/lib/credo/check/refactor/variable_rebinding.ex)

Thoughts?

silasjmatson commented on .credo.exs:
```diff
@@ -41,68 +50,94 @@
+        {Credo.Check.Refactor.ABCSize, false},
+        {Credo.Check.Refactor.AppendSingleItem, false},
+        {Credo.Check.Refactor.VariableRebinding, false},
+        {Credo.Check.Warning.MapGetUnsafePass, false},
```
I think we could opt into this as well.

zberkom committed:
:pencil: Updates per comments on PR

zberkom committed:
:sparkles: Add torch logo to readme

zberkom committed:
:white_check_mark: Fix credo error

zberkom commented:
@danielberkompas I just pushed three new commits that address much of your feedback.

### Mix tasks

1. `mix torch.install` 
   - Simplified to only install torch layout.
2. Removed `mix torch.gen.context` because I just don't see it being useful. Open to thoughts on that.
3. `mix torch.uninstall`
   - Simplified to look at configuration instead of passing arguments again
4. `mix torch.gen.html`
   - Now handles the adding and removing of the templates all at once.

#### Customization 

1. Updated `admin.html.eex` to `torch.html.eex`

#### Documentation 

1. Updated the Readme with logo
2. Added reminder to add link to `<nav></nav>`
3. I personally like the 🔥 symbols. We do something very similar on ignite.

#### Bugs

1. Added JS to fix flash message close button.

zberkom committed:
:boom: Delete torch context generator task
